tests/unit/LTI_Message_Launch_Test.php

```diff
@@ -0,0 +1,309 @@
+<?php
+
+namespace IMSGlobal\LTI\Tests\unit;
+
+use IMSGlobal\LTI\Cookie;
+use IMSGlobal\LTI\LTI_Message_Launch;
+use IMSGlobal\LTI\Tests\unit\helpers\InMemoryCache;
+
+class LTI_Message_Launch_Test extends TestBase {
+
+    private $publicKey;
+
+    private $validatorsLoaded = false;
+
+    public function setUp()
+    {
+        parent::setUp();        
+        $this->publicKey = file_get_contents(dirname(__FILE__) . '/fixtures/private.pub');
+    }
+
+    public function testNewInstance()
+    {
+        $this->assertInstanceOf(
+            'IMSGlobal\LTI\LTI_Message_Launch', 
+            LTI_Message_Launch::newInstance(
+                new DummyDatabase()
+            )
+        );
+    }    
+
+    public function testValidateStateWithInvalidStateThrowsException()
+    {
+        $this->setExpectedException('\IMSGlobal\LTI\LTI_Exception', 'State not found');
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder('\IMSGlobal\LTI\Cookie')
+            ->setMethods(['get_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->once())->method('get_cookie');
+
+        LTI_Message_Launch::newInstance(
+            new DummyDatabase(),
+            null,
+            $cookie
+        )->validate();
+    }
+
+    public function testValidateState()
+    {
+        $state = uniqid();
+        $this->setExpectedException('\IMSGlobal\LTI\LTI_Exception', 'Missing id_token');
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder('\IMSGlobal\LTI\Cookie')
+            ->setMethods(['get_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->atLeastOnce())->method('get_cookie')->with('lti1p3_' . $state)->willReturn($state);
+        LTI_Message_Launch::newInstance(
+            new DummyDatabase(),
+            null,
+            $cookie
+        )->validate(['state' => $state]);
+    }    
+
+    public function testValidateJWTFormat()
+    {
+        $jwt = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhYWEiLCJhdWQiOiIxMjM0NSJ9.MPkXNVeUj89bEt6nOC7KT3JW2KgY8Kks86xpL2uUatNLlzJk0HEd0BDq48D9ZSIGLpopkmV_pG1h_WHnXrX8Svi7-BnUQDz1u9Jomm57YlIz11XPM6FHoSVhUm1pdxLIYFbZu51vrJRSAYxm_zwiTnM_FLb3ZQZGAFyQ1rKGoevLy9Cyr_9gWyncoyXktz8NJ88nXCDf2MC2HsYRrDW48uXyFe6CtVhdHWqqLcNgw9LQNEtEaB_yU1aKkLSVKOX7LhCZ79rNx-06kG7LtgAhGGcOZYgtGGx4xaFJxJ7eBsvQBEkAyZbd4E6vZKHVEN-rQfreNUgTla_xdRfMdgDEpA';
```

Comment:
Is there anything about this particular JWT that is important? If so, should we construct it from data in the tests somewhere, rather than inlining it?

Reply:
Ah, it's badly signed.

Reply:
That's really just to stop the test. It might make more sense to stub out the other validation methods and not catch on an unrelated exception.

Reply:
Oh, nevermind, I can't stub them because they're all private